File: src/SpectracomDaemon/SpectracomStatus.py
import time

class SpectracomStatus(dict):
    '''Simple class holding a dictionary of status values for a Spectracom
       SecureSync time/frequency server
    '''
    
    ALLOWED_KEYS = ['_statusTime', '_hostName', '_hostResponding',
                    '_timeReference', '_ppsReference', '_ntpStratum',
                    '_inSync', '_oscType', '_oscState', '_tfom', '_maxTfom',
                    '_alarmStatusTime', '_majorAlarm', '_minorAlarm', 
                    '_alarmList', '_disciplineTime', '_dacValue',
                    '_ppsPhaseErr', '_freqErr', '_oscTemp', 'class_version']
    # Class version is required to play nicely with the C++ side using
    # Boost serialization. Keep this version number in sync with that in
    # SpectracomStatus.h.
    CLASS_VERSION = 0
    
    def __init__(self, hostName):
        # _statusTime: timestamp for the status information, seconds since
        # 1970-01-01 00:00:00 UTC
        # This time MUST be floating-point for compatibility with
        # our C++ SpectracomStatus class
        self['_statusTime'] = time.time()
        # _hostName: host name or IP address string for the Spectracom
        self['_hostName'] = hostName
        # _hostResponding: true iff the Spectracom is responding
        self['_hostResponding'] = False
        # _timeReference: string describing current time reference source
        self['_timeReference'] = '----'
        # _timeReference: string describing current 1PPS reference source
        self['_ppsReference'] = '----'
        # _ntpStratum: current NTP stratum
        self['_ntpStratum'] = 99
        # _inSync: true iff Spectracom is currently sync'ed to time and 1PPS
        # references
        self['_inSync'] = False
        # _oscType: type of oscillator installed in the Spectracom
        self['_oscType'] = 'Unknown'
        # _oscState: string describing the current oscillator state
        self['_oscState'] = 'Unknown'
        # _tfom: current time figure-of-merit
        self['_tfom'] = 99
        # _maxTfom: maximum TFOM where 1 PPS will be emitted
        self['_maxTfom'] = 99
        # _alarmStatusTime: timestamp of the latest alarm status report,
        # seconds since 1970-01-01 00:00:00 UTC.
        # This time MUST be floating-point for compatibility with
        # our C++ SpectracomStatus class
        self['_alarmStatusTime'] = 0.0
        # _majorAlarm: true iff a major alarm is active
        self['_majorAlarm'] = False
        # _minorAlarm: true iff a minor alarm is active
        self['_minorAlarm'] = False
        # _alarmList: string with list of currently active alarms separated
        # by newlines
        self['_alarmList'] = ''
        # _disciplineTime: timestamp of the latest discipline report, seconds
        # since 1970-01-01 00:00:00 UTC
        # This is the time which applies to _dacValue, _ppsPhaseErr, _freqErr,
        # and _oscTemp.
        # This time MUST be floating-point for compatibility with
        # our C++ SpectracomStatus class
        self['_disciplineTime'] = 0.0
        # _dacValue: oscillator discipline DAC value
        self['_dacValue'] = 0
        # _ppsPhaseErr: 1 PPS phase error, ns
        self['_ppsPhaseErr'] = 9999999
        # _freqErr: latest frequency error
        self['_freqErr'] = 999.e9
        # _oscTemp: oscillator temperature
        self['_oscTemp'] = 0.0
        # C++ SpectracomStatus class version, needed for interoperability
        # with Boost serialization
        self['class_version'] = self.CLASS_VERSION
        
        # Verify that all valid keys have been initialized
        for k in self.ALLOWED_KEYS:
            if not self.has_key(k):
                raise RuntimeError, 'Value for key %s was not initialized' % (key)
        
    def __setitem__(self, key, value):
        if not key in self.ALLOWED_KEYS:
            raise ValueError, 'Bad key "%s" for SpectracomStatus' % (key)
        return dict.__setitem__(self, key, value)
    
    def __getitem__(self, key):
        if not key in self.ALLOWED_KEYS:
            raise ValueError, 'Bad key "%s" for SpectracomStatus' % (key)
        return dict.__getitem__(self, key)

File: src/SpectracomDaemon/SpectracomStatus.cpp
// *=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=* 
// ** Copyright UCAR (c) 1990 - 2016                                         
// ** University Corporation for Atmospheric Research (UCAR)                 
// ** National Center for Atmospheric Research (NCAR)                        
// ** Boulder, Colorado, USA                                                 
// ** BSD licence applies - redistribution and use in source and binary      
// ** forms, with or without modification, are permitted provided that       
// ** the following conditions are met:                                      
// ** 1) If the software is modified to produce derivative works,            
// ** such modified software should be clearly marked, so as not             
// ** to confuse it with the version available from UCAR.                    
// ** 2) Redistributions of source code must retain the above copyright      
// ** notice, this list of conditions and the following disclaimer.          
// ** 3) Redistributions in binary form must reproduce the above copyright   
// ** notice, this list of conditions and the following disclaimer in the    
// ** documentation and/or other materials provided with the distribution.   
// ** 4) Neither the name of UCAR nor the names of its contributors,         
// ** if any, may be used to endorse or promote products derived from        
// ** this software without specific prior written permission.               
// ** DISCLAIMER: THIS SOFTWARE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS  
// ** OR IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED      
// ** WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.    
// *=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=* 
/*
 * SpectracomStatus.cpp
 *
 *  Created on: Aug 31, 2012
 *      Author: burghart
 */

#include <cstdlib>
#include <Archive_xmlrpc_c.h>
#include <logx/Logging.h>
#include <SpectracomDaemon/SpectracomStatus.h>

LOGGING("SpectracomStatus")

SpectracomStatus::SpectracomStatus() :
    _statusTime(time(0)),
    _hostName("Unknown"),
    _hostResponding(false),
    _timeReference("----"),
    _ppsReference("----"),
    _ntpStratum(999),
    _inSync(false),
    _oscType("Unknown"),
    _oscState("Unknown"),
    _tfom(999),
    _maxTfom(999),
    _alarmStatusTime(0.0),
    _majorAlarm(false),
    _minorAlarm(false),
    _alarmList(),
    _disciplineTime(0.0),
    _dacValue(0),
    _ppsPhaseErr(0),
    _freqErr(1.0),
    _oscTemp(0.0) {
}

SpectracomStatus::SpectracomStatus(xmlrpc_c::value_struct & statusDict):
    _statusTime(time(0)),
    _hostName(),
    _hostResponding(false),
    _timeReference("----"),
    _ppsReference("----"),
    _ntpStratum(999),
    _inSync(false),
    _oscType("Unknown"),
    _oscState("Unknown"),
    _tfom(999),
    _maxTfom(999),
    _alarmStatusTime(0.0),
    _majorAlarm(false),
    _minorAlarm(false),
    _alarmList(),
    _disciplineTime(0.0),
    _dacValue(0),
    _ppsPhaseErr(0),
    _freqErr(1.0),
    _oscTemp(0.0) {
    // Create an input archiver wrapper around the xmlrpc_c::value_struct
    // dictionary, and use serialize() to populate our members from its content.
    Iarchive_xmlrpc_c iar(statusDict);
    iar >> *this;
}

SpectracomStatus::~SpectracomStatus() {
}

xmlrpc_c::value_struct
SpectracomStatus::toXmlRpcValue() const {
    std::map<std::string, xmlrpc_c::value> statusDict;
    // Clone ourself to a non-const instance
    SpectracomStatus clone(*this);
    // Stuff our content into the statusDict, i.e., _serialize() to an
    // output archiver wrapped around the statusDict.
    Oarchive_xmlrpc_c oar(statusDict);
    oar << clone;
    // Finally, return the statusDict
    return(xmlrpc_c::value_struct(statusDict));
}

int
SpectracomStatus::syncSimpleStatus() const {
    // Currently, we indicate only minor problems or none for sync state
    return(_inSync ? 0 : 1);
}

int
SpectracomStatus::alarmsSimpleStatus() const {
    if (_majorAlarm) {
        return 2;
    } else if (_minorAlarm) {
        return 1;
    }
    return 0;
}

int
SpectracomStatus::tfomSimpleStatus() const {
    // Currently, we indicate only minor problems or none for TFOM state
    return((_tfom <= _maxTfom) ? 0 : 1);
}

int
SpectracomStatus::oscSimpleStatus() const {
    // Currently, we indicate only minor problems or none for oscillator state
    return((_oscState == "Trk/Lock") ? 0 : 1);
}

int
SpectracomStatus::simpleStatus() const {
    // Major problem if the host is not responding
    int status = _hostResponding ? 0 : 2;
    // Change overall status if we get worse status for sync, alarms, TFOM,
    // or the oscillator
    if (syncSimpleStatus() > status) {
        status = syncSimpleStatus();
    }
    if (alarmsSimpleStatus() > status) {
        status = alarmsSimpleStatus();
    }
    if (tfomSimpleStatus() > status) {
        status = tfomSimpleStatus();
    }
    if (oscSimpleStatus() > status) {
        status = oscSimpleStatus();
    }

    return(status);
}
